src/compiler/commandLineParser.ts

```diff
@@ -1212,94 +1511,223 @@ namespace ts {
+        // typingOptions has been deprecated and is only supported for backward compatibility purposes.
+        // It should be removed in future releases - use typeAcquisition instead.
+        const typeAcquisition = convertTypeAcquisitionFromJsonWorker(json["typeAcquisition"] || json["typingOptions"], basePath, errors, configFileName);
+        json.compileOnSave = convertCompileOnSaveOptionFromJson(json, basePath, errors);
+        let extendedConfigPath: Path;
 
         if (json.extends) {
-            // copy the resolution stack so it is never reused between branches in potential diamond-problem scenarios.
-            resolutionStack = resolutionStack.concat([resolvedPath]);
-            const base = getExtendedConfig(json.extends, host, basePath, getCanonicalFileName, resolutionStack, errors);
-            if (base) {
-                include = include || base.include;
-                exclude = exclude || base.exclude;
-                files = files || base.files;
-                if (compileOnSave === undefined) {
-                    compileOnSave = base.compileOnSave;
-                }
-                options = assign({}, base.options, options);
+            if (typeof json.extends !== "string") {
+                errors.push(createCompilerDiagnostic(Diagnostics.Compiler_option_0_requires_a_value_of_type_1, "extends", "string"));
+            }
+            else {
+                extendedConfigPath = getExtendsConfigPath(json.extends, host, basePath, getCanonicalFileName, errors, createCompilerDiagnostic);
             }
         }
-
-        return { include, exclude, files, options, compileOnSave };
+        return { raw: json, options, typeAcquisition, extendedConfigPath };
     }
 
-    function getExtendedConfig(
-            extended: any, // Usually a string.
-            host: ts.ParseConfigHost,
-            basePath: string,
-            getCanonicalFileName: (fileName: string) => string,
-            resolutionStack: Path[],
-            errors: Diagnostic[],
-            ): ParsedTsconfig | undefined {
-        if (typeof extended !== "string") {
-            errors.push(createCompilerDiagnostic(Diagnostics.Compiler_option_0_requires_a_value_of_type_1, "extends", "string"));
-            return undefined;
+    function parseOwnConfigOfJsonSourceFile(
+        sourceFile: JsonSourceFile,
+        host: ParseConfigHost,
+        basePath: string,
+        getCanonicalFileName: (fileName: string) => string,
+        configFileName: string,
+        errors: Diagnostic[]
+    ): ParsedTsconfig {
+        const options = getDefaultCompilerOptions(configFileName);
+        let typeAcquisition: TypeAcquisition, typingOptionstypeAcquisition: TypeAcquisition;
+        let extendedConfigPath: Path;
+
+        const optionsIterator: JsonConversionNotifier = {
+            onSetValidOptionKeyValueInParent(parentOption: string, option: CommandLineOption, value: CompilerOptionsValue) {

```

Comment:
Type annotations on parameters are not necessary, those should be inherited from `JsonConversionNotifier`.